### ISSUE TITLE
FireStore node saving STRING as NUMBER

### ISSUE THREAD

**Issue by rahil320** [n8n-io/n8n#11413]
### Bug Description

I have created this workflow where I use the **SET node** to create new values (variables) and then use **Google Cloud Firestore** node to store data in a collection. In the SET node I made few variables which is declared as STRING but has a value of number like 9999. When I execute this workflow, n8n creates field in the document but changes the datatype in firestore to number. When I use double quotes to enclose variable value in SET node, then the workflow add additional quotes in the firestore document field.

### To Reproduce

1. Create a Manual Trigger, then SET node and then Google firestore node
2. Connect Manual trigger to SET node
3. Edit the SET node and use Manual Mapping to create a STRING field.
4. Set the value of this STRING field to any numeric data like 9999
5. Now set up the Google firestore as follows (or similar)
     Set **Resource** to **Document**
     **Operation** to **Create or Update**
     **Project** and **Database** to anything suitable
     **Collection** to any suitable collection name
     **Update Key** to any document ID (in firebase)
     **Columns/Attributes** to the names of fields used in **SET** node
6. Now execute the n8n workflow and fields would be generated in Firestore collection

![step1](https://github.com/user-attachments/assets/0685fe97-36eb-4b64-a01f-82848861a2a8)
![Step2](https://github.com/user-attachments/assets/069b2ee1-ea92-4f50-9a20-efd4a2ff485a)
![Step3](https://github.com/user-attachments/assets/1add442e-2998-4f85-bcfc-ac4f56bbf58a)
![Step4](https://github.com/user-attachments/assets/54c20eb3-7370-4107-b15c-5fe93ae32eb2)

### Expected behavior

Since I have set the datatype as STRING in the SET node so the datatype in Google firestore should also remain the STRING but somehow it gets converted to NUMBER which causes the leading zeros to disappear.

### Operating System

Ubuntu 24.04.1 LTS

### n8n Version

Version 1.64.3

### Node.js Version

v20.18.0

### Database

SQLite (default)

### Execution mode

main (default)

**Comment by Joffcom** [comment]
Hey @rahil320,

We have created an internal ticket to look into this which we will be tracking as "GHC-350"

**Comment by Heifarabuval** [comment]
**Hello,**  

I am facing the same issue with the Firestore node in n8n. Even though I explicitly define certain variables as **STRING** in the SET node, Firestore still interprets them as **NUMBER** when they contain numeric values.  

### **Workarounds I have tested:**  
1. **Prefixing numeric values with a string**  
   - I had to add a prefix (e.g., `"str_" + value`) to force Firestore to retain the STRING type.  
   - Without this workaround, Firestore automatically converts numeric strings into NUMBER.  

2. **Empty strings (`""`) are stored as `null`**  
   - Firestore seems to treat empty strings as `null`, which can cause issues in certain cases.  
   - A temporary fix is to replace `""` with `"N/A"` or `"empty"`, but this is not ideal.  

### **Waiting for a solution**  
It would be great to have an **option to enforce field types** in Firestore, preventing automatic type conversions. This would eliminate the need for manual prefixing or handling unintended `null` values.  

Looking forward to any updates or a possible fix! 😊

**Comment by frankjoshua** [comment]
I also have an issue with the Firestore node converting types that seems to be related to this. If I try to create a document with a integer field of 0 it will save that field as null. Positive numbers work fine. I did not test with negative.

**Comment by Heifarabuval** [comment]
The problem comes from the `jsonToDocument` function that turns our input data into Firestore format.

Right now, if you give it a string that looks like a number (like `"09871234"`), it treats it as a number because of this check:

```ts
if (value !== '' && !isNaN(value as number)) {
    if (value.toString().indexOf('.') !== -1) {
        return { doubleValue: value };
    } else {
        return { integerValue: value };
    }
}
```

That means Firestore gets an `integerValue` instead of a `stringValue`.

Also, empty strings (`""`) end up as `nullValue`, which is annoying.

### So where to intervene ?

We need to change `jsonToDocument`:

* Add a way to force values to stay strings, even if they look like numbers.
* Change how empty strings are handled so they stay empty strings in Firestore, not null.

We could add an option to the function or the node to control this behavior, so users can decide if they want strict typing or not.

Basically, the fix is here, before data goes to Firestore.

**Comment by hchafiatheal** [comment]
also having this issue, hopefully can be addressed soon.

**Comment by mufsalup** [comment]
I have the very same problem here! I clearly receive Strings, modifying Strings in n8n, sending them (as Strings?) retrieving numbers.

**Comment by Heifarabuval** [comment]
I've been looking into `jsonToDocument` in `GenericFunctions.ts` and found the issue.

The current check `!isNaN(value as number)` is too aggressive. It forces any string composed of digits (like "0123" or "9999") to be cast as a Firestore number, effectively overriding the explicit types set in previous nodes (like the SET node).

We should simply respect the upstream JS type. If it's strictly a `number`, map it to a numeric Firestore type. If it's a `string`, let it be a string.

Here is a fix that relies on `typeof` instead of `isNaN`:

```typescript
// packages/nodes-base/nodes/Google/Firebase/CloudFirestore/GenericFunctions.ts

export function jsonToDocument(value: string | number | IDataObject | IDataObject[]): IDataObject {
    if (value === 'true' || value === 'false' || typeof value === 'boolean') {
        return { booleanValue: value };
    } else if (value === null) {
        return { nullValue: null };
        
    // CHANGED HERE: 
    // Old logic used !isNaN which was catching numeric strings
    // } else if (value !== '' && !isNaN(value as number)) {
    
    // New logic: strict type check
    } else if (typeof value === 'number') {
        if (value.toString().indexOf('.') !== -1) {
            return { doubleValue: value };
        } else {
            return { integerValue: value };
        }
    } else if (isValidDate(value as string)) {
        const date = new Date(Date.parse(value as string));
        return { timestampValue: date.toISOString() };
    } else if (typeof value === 'string') {
        return { stringValue: value };
    // ... rest of the function

```

This simple change solves the issue:

1. "9999" stays a `stringValue` because it fails the strict number check.
2. Empty strings `""` are correctly handled as strings instead of ending up as `null` or `0`.

Let me know if this makes sense, I can open a PR.
